### PR TITLE
Aws api memory leak fix

### PR DESCRIPTION
**Pull request by hemslo**
This change fixed memory leak in AWS API.

1. bump boto3 version so `close` is available
2. refactored `lru_cache` to merge methods for clients
3. collect all clients in `AWSApi` so they can be closed
4. update all usages of `AWSApi` to run `cleanup`

APPSRE-4141

### CODE REVIEW

File: reconcile/ocp_release_mirror.py [hunk 1]
```diff
@@ -363,3 +367,6 @@ def run(dry_run: bool) -> None:
         except OcpReleaseMirrorError as details:
             LOG.error(str(details))
             sys.exit(ExitCodes.ERROR)
+        finally:
+            if quay_mirror is not None:
```
Comment (chassing) [inline reconcile/ocp_release_mirror.py:371]: Why this `if` statement?

**Review by chassing — COMMENTED** [review]
I love context managers. Very nice refactoring. ❤️

File: reconcile/utils/saasherder.py [hunk 2]
```diff
@@ -209,6 +209,16 @@ def __init__(
...
             self.state = init_state(integration=self.integration)
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc):
+        self.cleanup()
+
+    def cleanup(self):
+        if self.state:
```
[the remark this answers is not in the document]
Comment (hemslo) [reply reconcile/utils/saasherder.py:219]: cool, I'll check attr exist and scan all conditional cleanup, thanks!

File: reconcile/ocp_release_mirror.py [hunk 3]
```diff
@@ -191,6 +191,10 @@ def __init__(self, dry_run: bool, instance: OcpReleaseMirrorV1):
         # Initiate channel groups
         self.channel_groups = instance.mirror_channels
 
+    def cleanup(self):
+        if self.aws_cli:
```
[the remark this answers is not in the document]
Comment (hemslo) [reply reconcile/ocp_release_mirror.py:195]: seems not needed, I'll remove it and check all conditional cleanup

File: reconcile/ocp_release_mirror.py (hunk 1 above)
Comment (hemslo) [reply reconcile/ocp_release_mirror.py:371]: not needed, I'll remove it, thanks!

Comment (hemslo) [reply reconcile/ocp_release_mirror.py:371]: this is still needed, but need to be done differently, `__init__` can raise error, so need to check if `quay_mirror` exist

**Review by chassing — APPROVED** [review]
LGTM

**Pull request merged**